bug: CloudFormation dynamic reference regression 4.7 to 4.8

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Current Behavior

I am creating a stack with the following CloudFormation template:

```yaml
AWSTemplateFormatVersion: 2010-09-09
Transform: AWS::Serverless-2016-10-31

Parameters:
  Profile:
    Type: String
    Description: The stage to publish on deploy
    AllowedValues:
      - prod
      - stag
      - test

  LambdaAlias:
    Type: String
    Description: The Lambda function alias to publish on deploy
    AllowedValues:
      - live
      - latest

  BuildId:
    Type: String
    Description: The build identifier
    AllowedPattern: "^[0-9A-Za-z]+$"
    
  CognitoUserPoolId:
    Type: String
    Description: The Cognito User Pool ID to use for the post confirmation hook
    AllowedPattern: '^[a-z]{2}-[a-z]+-\d+_[A-Za-z0-9]{8,}$'
    
  Bump:
    Type: String
    Description: An arbitrary parameter to force CloudFormation to update the stack
    Default: "0"
    AllowedPattern: "^[0-9A-Za-z]+$"
    
Conditions:
  IsTesting: !Equals [ !Ref Profile, "test" ]
  
Mappings:
  LambdaConfiguration:
    prod:
      StripeSecretName: foobar/prod/stripe
      ApiKeySigningSecretName: /foobar/prod/signing
      FreeAwsPlanParameterName: /foobar/prod/plans/free/aws
      SmallAwsPlanParameterName: /foobar/prod/plans/small/aws
      MediumAwsPlanParameterName: /foobar/prod/plans/medium/aws
      LargeAwsPlanParameterName: /foobar/prod/plans/large/aws
    stag:
      StripeSecretName: foobar/stag/stripe
      ApiKeySigningSecretName: /foobar/stag/signing
      FreeAwsPlanParameterName: /foobar/stag/plans/free/aws
      SmallAwsPlanParameterName: /foobar/stag/plans/small/aws
      MediumAwsPlanParameterName: /foobar/stag/plans/medium/aws
      LargeAwsPlanParameterName: /foobar/stag/plans/large/aws
    test:
      StripeSecretName: foobar/test/stripe
      ApiKeySigningSecretName: /foobar/test/signing
      FreeAwsPlanParameterName: /foobar/test/plans/free/aws
      SmallAwsPlanParameterName: /foobar/test/plans/small/aws
      MediumAwsPlanParameterName: /foobar/test/plans/medium/aws
      LargeAwsPlanParameterName: /foobar/test/plans/large/aws

Resources:
  # NOTE: This Lambda Function requires some clickops.
  #
  # You must create the following Secrets Manager secrets manually:
  # - foobar/prod/stripe
  # - foobar/stag/stripe
  #
  # You must create the following SSM Parameters manually:
  # - /foobar/prod/plans/free/aws
  # - /foobar/prod/plans/small/aws
  # - /foobar/prod/plans/medium/aws
  # - /foobar/prod/plans/large/aws
  # - /foobar/stag/plans/free/aws
  # - /foobar/stag/plans/small/aws
  # - /foobar/stag/plans/medium/aws
  # - /foobar/stag/plans/large/aws  
  
  LambdaFunction:
    Type: AWS::Serverless::Function
    Properties:    
      Architectures:
        - x86_64
      AutoPublishAlias: !Ref LambdaAlias
      # pragma package lambda-cognito-post-confirmation-hook.jar
      CodeUri: "target/lambda-cognito-post-confirmation-hook.jar"
      Description: Prepares new cognito accounts for Stripe usage
      Environment:
        Variables:
          PROFILE: !Ref Profile
          API_KEY_SIGNING_SECRET_NAME: !FindInMap [LambdaConfiguration, !Ref Profile, ApiKeySigningSecretName]
          STRIPE_SECRET_NAME: !FindInMap [LambdaConfiguration, !Ref Profile, StripeSecretName]
          COGNITO_USER_POOL_ID: !Ref CognitoUserPoolId
          AWS_FREE_USAGE_PLAN_ID:
            Fn::Sub:
              - "{{resolve:ssm:${ParameterName}}}"
              - ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, FreeAwsPlanParameterName]
          AWS_SMALL_USAGE_PLAN_ID:
            Fn::Sub:
              - "{{resolve:ssm:${ParameterName}}}"
              - ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, SmallAwsPlanParameterName]
          AWS_MEDIUM_USAGE_PLAN_ID:
            Fn::Sub:
              - "{{resolve:ssm:${ParameterName}}}"
              - ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, MediumAwsPlanParameterName]
          AWS_LARGE_USAGE_PLAN_ID:
            Fn::Sub:
              - "{{resolve:ssm:${ParameterName}}}"
              - ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, LargeAwsPlanParameterName]
          BUMP: !Ref Bump
      FunctionName: !Sub "foobar-${Profile}-stripe-cognito-post-confirmation-hook"
      Handler: io.foobar.portal.backend.lambda.cognito.hook.postconfirmation.LambdaFunction
      MemorySize: 4096
      PackageType: Zip
      Policies:
        - AWSLambdaBasicExecutionRole
        - AWSXRayDaemonWriteAccess
        - Statement:
            - Effect: Allow
              Action:
                - secretsmanager:GetSecretValue
              Resource:
                - Fn::Sub:
                  - "arn:aws:secretsmanager:${Region}:${AccountId}:secret:${SecretName}-*"
                  - Region: !Ref AWS::Region
                    AccountId: !Ref AWS::AccountId
                    SecretName: !FindInMap [LambdaConfiguration, !Ref Profile, StripeSecretName]
                - Fn::Sub:
                  - "arn:aws:secretsmanager:${Region}:${AccountId}:secret:${SecretName}-*"
                  - Region: !Ref AWS::Region
                    AccountId: !Ref AWS::AccountId
                    SecretName: !FindInMap [LambdaConfiguration, !Ref Profile, ApiKeySigningSecretName]
        - Statement:
            - Effect: Allow
              Action:
                - ssm:GetParameter
              Resource:
                - !Sub "arn:aws:ssm:${AWS::Region}:${AWS::AccountId}:parameter/foobar/${Profile}/*"
        - Statement:
            - Effect: Allow
              Action:
                - cognito-idp:AdminGetUser
                - cognito-idp:AdminUpdateUserAttributes
              Resource:
                - !Sub "arn:aws:cognito-idp:${AWS::Region}:${AWS::AccountId}:userpool/${CognitoUserPoolId}"
        - Statement:
            # We need to manage API keys
            - Effect: Allow
              Action:
                # List all API keys
                - apigateway:GET
                # Create new API keys
                - apigateway:POST
              Resource:
                - !Sub "arn:aws:apigateway:${AWS::Region}::/apikeys"
            # We need to create API keys for the free usage plan
            - Effect: Allow
              Action:
                - apigateway:POST
              Resource:
                - Fn::Sub:
                  - "arn:aws:apigateway:${Region}::/usageplans/{{resolve:ssm:${ParameterName}}}/keys"
                  - Region: !Ref AWS::Region
                    ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, FreeAwsPlanParameterName]
            # We need to delete API keys for all usage plans, just in case
            - Effect: Allow
              Action:
                - apigateway:DELETE
              Resource:
                - Fn::Sub:
                  - "arn:aws:apigateway:${Region}::/usageplans/{{resolve:ssm:${ParameterName}}}/keys/*"
                  - Region: !Ref AWS::Region
                    ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, FreeAwsPlanParameterName]
                - Fn::Sub:
                  - "arn:aws:apigateway:${Region}::/usageplans/{{resolve:ssm:${ParameterName}}}/keys/*"
                  - Region: !Ref AWS::Region
                    ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, SmallAwsPlanParameterName]
                - Fn::Sub:
                  - "arn:aws:apigateway:${Region}::/usageplans/{{resolve:ssm:${ParameterName}}}/keys/*"
                  - Region: !Ref AWS::Region
                    ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, MediumAwsPlanParameterName]
                - Fn::Sub:
                  - "arn:aws:apigateway:${Region}::/usageplans/{{resolve:ssm:${ParameterName}}}/keys/*"
                  - Region: !Ref AWS::Region
                    ParameterName: !FindInMap [LambdaConfiguration, !Ref Profile, LargeAwsPlanParameterName]
      Runtime: java21
      RuntimeManagementConfig:
        UpdateRuntimeOn: FunctionUpdate
      # This lambda function has no sophisticated state, so TURN OFF SnapStart!
      # SnapStart:
      #   ApplyOn: PublishedVersions
      Tags:
        "aleph0": "true"
        "aleph0:scope": !Ref Profile
        "aleph0:product": foobar
      Timeout: 30
      Tracing: !If [IsTesting, !Ref AWS::NoValue, "Active"]
      VersionDescription: !Ref BuildId

Outputs:
  LambdaFunction:
    Value: !GetAtt LambdaFunction.Arn
    Description: The ARN of the Stripe Cognito Post Confirmation Hook Lambda function
    Export:
      Name: !Sub "foobar-${Profile}-stripe-cognito-post-confirmation-hook"
```

### Expected Behavior

Using LocalStack v4.7, the stack creates as expected. Using LocalStack v4.8, the create fails. Oddly, after I wait for the stack's status to settle, the stack reports status `CREATE_COMPLETE`. However, the following events appear in the event history, sorted by timestamp descending, from my logs:

```
[22:59:58.738] [ERROR] [main] LambdaFunctionITBase - Stack in state CREATE_COMPLETE, but at least one resource actually failed to create: LambdaFunction = An error occurred (ParameterNotFound) when calling the GetParameter operation: Parameter ${ParameterName} not found.
[22:59:58.748] [ERROR] [main] LambdaFunctionITBase - Stack event: 7bee3577-8f4b-4df6-bfe4-6ce980864900 2025-09-19T03:59:57.671973Z my-test-stack AWS::CloudFormation::Stack CREATE_COMPLETE null
[22:59:58.748] [ERROR] [main] LambdaFunctionITBase - Stack event: 58d9b7ef-1576-4ab9-bedc-af1233acd390 2025-09-19T03:59:57.671964Z my-test-stack AWS::CloudFormation::Stack ROLLBACK_COMPLETE null
[22:59:58.748] [ERROR] [main] LambdaFunctionITBase - Stack event: 0762417c-969f-4ec2-89c2-c4f87bb08ace 2025-09-19T03:59:57.671937Z my-test-stack AWS::CloudFormation::Stack CREATE_FAILED null
[22:59:58.748] [ERROR] [main] LambdaFunctionITBase - Stack event: f84893e9-29e5-451f-beeb-57d82691ca55 2025-09-19T03:59:57.671929Z LambdaFunction AWS::Lambda::Function CREATE_FAILED An error occurred (ParameterNotFound) when calling the GetParameter operation: Parameter ${ParameterName} not found.
[22:59:58.748] [ERROR] [main] LambdaFunctionITBase - Stack event: 596e56f8-cfb9-4a9a-bff9-773fad8a1319 2025-09-19T03:59:57.665751Z my-test-stack AWS::CloudFormation::Stack CREATE_IN_PROGRESS null
```

Note the presence of a `CREATE_FAILED` event on the `LambdaFunction` resource.



### How are you starting LocalStack?

Custom (please describe below)

### Steps To Reproduce

#### How are you starting localstack (e.g., `bin/localstack` command, arguments, or `docker-compose.yml`)

I'm starting LocalStack inside my Java JUnit5 tests using this code:

```java
  @Container
  @SuppressWarnings("resource")
  public static LocalStackContainer localstack =
      new LocalStackContainer(DockerImageName.parse("localstack/localstack:4.7"))
          .withServices(LocalStackContainer.Service.CLOUDFORMATION, LocalStackContainer.Service.S3,
              LocalStackContainer.Service.LAMBDA, LocalStackContainer.Service.IAM,
              LocalStackContainer.Service.SSM, LocalStackContainer.Service.SECRETSMANAGER)
          // Let's get some debug output if things go wrong
          // .withEnv("DEBUG", "1")
          // Increase the Lambda timeout to 5 minutes to avoid intermittent failures
          .withEnv("LAMBDA_RUNTIME_ENVIRONMENT_TIMEOUT", "300")
          // Use the "local" executor to avoid intermittent failures
          // .withEnv("LAMBDA_RUNTIME_EXECUTOR", "local")
          // Give LocalStack enough time to start up
          .withStartupTimeout(Duration.ofMinutes(1L));
```

#### Client commands (e.g., AWS SDK code snippet, or sequence of "awslocal" commands)

I am creating the stack using the AWS Java SDK v2 CloudFormation client.



### Environment

```markdown
- OS: MacOS 15.6.1 (local), Ubuntu 24.04.3 LTS (GitHub Actions) both
- LocalStack:
  LocalStack version: 4.7, 4.8
  LocalStack Docker image sha: 4.8: a6d609a1714f, 4.7: ad4f76a02108
  LocalStack build date:
  LocalStack build git hash:
```

### Anything else?

My best guess about what's going on here is a change in "order of operations" for dynamic reference resolution. I noticed some PRs that were merged between 4.7 and 4.8 that seem relevant. I have not run the issue to ground, but these came up in my search for an existing issue, so I'm providing them here in case it's useful information:

* https://github.com/localstack/localstack/pull/12934
* https://github.com/localstack/localstack/pull/12965

I use these parameters in tests for creating the template:

    Profile: test
    LambdaAlias: test
    BuildId: 1234
    CognitoUserPoolId: us-east-1_GoodUserPoolId

These steps are required to "prepare" for creating that template, in Java code, from my tests:

```java
    ssm.putParameter(b -> b.name("/foobar/test/plans/free/aws").type(ParameterType.STRING)
        .value("free.aws"));
    ssm.putParameter(b -> b.name("/foobar/test/plans/small/aws").type(ParameterType.STRING)
        .value("small.aws"));
    ssm.putParameter(b -> b.name("/foobar/test/plans/medium/aws").type(ParameterType.STRING)
        .value("medium.aws"));
    ssm.putParameter(b -> b.name("/foobar/test/plans/large/aws").type(ParameterType.STRING)
        .value("large.aws"));

    secretsManager.createSecret(
        b -> b.name("foobar/test/stripe").secretString("{\"apiKey\":\"S00p3rS3cret!\"}"));
    secretsManager.createSecret(
        b -> b.name("/foobar/test/signing").secretString("{\"hex\":\"1234\"}"));
```

--- a/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
@@ -435,19 +435,13 @@ class ChangeSetModelPreproc(ChangeSetModelVisitor):
             return value
 
         if dynamic_ref := extract_dynamic_reference(value):
-            try:
-                new_value = perform_dynamic_reference_lookup(
-                    reference=dynamic_ref,
-                    account_id=self._change_set.account_id,
-                    region_name=self._change_set.region_name,
-                )
-                if new_value:
-                    return REGEX_DYNAMIC_REF.sub(new_value, value)
-            except Exception:
-                # we cannot look this reference up right now, however that might be because we
-                # are visiting the arguments to an intrinsic function that will eventually create
-                # a full reference.
-                pass
+            new_value = perform_dynamic_reference_lookup(
+                reference=dynamic_ref,
+                account_id=self._change_set.account_id,
+                region_name=self._change_set.region_name,
+            )
+            if new_value:
+                return REGEX_DYNAMIC_REF.sub(new_value, value)
 
         return value
 

--- a/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
@@ -434,7 +434,7 @@ class ChangeSetModelPreproc(ChangeSetModelVisitor):
         if not isinstance(value, str):
             return value
         if isinstance(value, str) and (dynamic_ref_match := REGEX_DYNAMIC_REF.search(value)):
-            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2], dynamic_ref_match[3])
+            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
             try:
                 new_value = perform_dynamic_reference_lookup(
                     reference=ref,

--- a/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
@@ -45,7 +45,8 @@ from localstack.services.cloudformation.engine.v2.change_set_model_visitor impor
     ChangeSetModelVisitor,
 )
 from localstack.services.cloudformation.engine.v2.resolving import (
-    extract_dynamic_reference,
+    REGEX_DYNAMIC_REF,
+    DynamicReference,
     perform_dynamic_reference_lookup,
 )
 from localstack.services.cloudformation.engine.validations import ValidationError
@@ -432,14 +433,21 @@ class ChangeSetModelPreproc(ChangeSetModelVisitor):
     def _perform_dynamic_replacements(self, value: _T) -> _T:
         if not isinstance(value, str):
             return value
-        if dynamic_ref := extract_dynamic_reference(value):
-            new_value = perform_dynamic_reference_lookup(
-                reference=dynamic_ref,
-                account_id=self._change_set.account_id,
-                region_name=self._change_set.region_name,
-            )
-            if new_value:
-                return new_value
+        if isinstance(value, str) and (dynamic_ref_match := REGEX_DYNAMIC_REF.search(value)):
+            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
+            try:
+                new_value = perform_dynamic_reference_lookup(
+                    reference=ref,
+                    account_id=self._change_set.account_id,
+                    region_name=self._change_set.region_name,
+                )
+                if new_value:
+                    return REGEX_DYNAMIC_REF.sub(new_value, value)
+            except Exception:
+                # we cannot look this reference up right now, however that might be because we
+                # are visiting the arguments to an intrinsic function that will eventually create
+                # a full reference.
+                pass
 
         return value
 

--- a/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
@@ -434,7 +434,7 @@ class ChangeSetModelPreproc(ChangeSetModelVisitor):
         if not isinstance(value, str):
             return value
         if isinstance(value, str) and (dynamic_ref_match := REGEX_DYNAMIC_REF.search(value)):
-            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
+            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2], dynamic_ref_match[3])
             try:
                 new_value = perform_dynamic_reference_lookup(
                     reference=ref,

--- a/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/change_set_model_preproc.py
@@ -46,7 +46,7 @@ from localstack.services.cloudformation.engine.v2.change_set_model_visitor impor
 )
 from localstack.services.cloudformation.engine.v2.resolving import (
     REGEX_DYNAMIC_REF,
-    DynamicReference,
+    extract_dynamic_reference,
     perform_dynamic_reference_lookup,
 )
 from localstack.services.cloudformation.engine.validations import ValidationError
@@ -433,11 +433,11 @@ class ChangeSetModelPreproc(ChangeSetModelVisitor):
     def _perform_dynamic_replacements(self, value: _T) -> _T:
         if not isinstance(value, str):
             return value
-        if isinstance(value, str) and (dynamic_ref_match := REGEX_DYNAMIC_REF.search(value)):
-            ref = DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
+
+        if dynamic_ref := extract_dynamic_reference(value):
             try:
                 new_value = perform_dynamic_reference_lookup(
-                    reference=ref,
+                    reference=dynamic_ref,
                     account_id=self._change_set.account_id,
                     region_name=self._change_set.region_name,
                 )

--- a/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
@@ -9,21 +9,15 @@ from localstack.aws.connect import connect_to
 
 LOG = logging.getLogger(__name__)
 
-REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([a-zA-Z0-9_.\-/]+)(?::(\d+))?}}")
+# CloudFormation allows using dynamic references in `Fn::Sub` expressions, so we must make sure
+# we don't capture the parameter usage by excluding ${} characters
+REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([^${}]+)}}")
 
 
 @dataclass
 class DynamicReference:
     service_name: str
     reference_key: str
-    version: str | None = None
-
-    @property
-    def versioned_name(self) -> str:
-        if self.version:
-            return f"{self.reference_key}:{self.version}"
-        else:
-            return self.reference_key
 
 
 def perform_dynamic_reference_lookup(
@@ -38,7 +32,7 @@ def perform_dynamic_reference_lookup(
     if reference.service_name == "ssm":
         ssm_client = connect_to(aws_access_key_id=account_id, region_name=region_name).ssm
         try:
-            return ssm_client.get_parameter(Name=reference.versioned_name)["Parameter"]["Value"]
+            return ssm_client.get_parameter(Name=reference.reference_key)["Parameter"]["Value"]
         except ClientError as e:
             LOG.error("client error accessing SSM parameter '%s': %s", reference.reference_key, e)
             raise

--- a/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
@@ -9,7 +9,7 @@ from localstack.aws.connect import connect_to
 
 LOG = logging.getLogger(__name__)
 
-REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):(.+)}}")
+REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([a-zA-Z0-9_.\-/]+)}}")
 
 
 @dataclass

--- a/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
@@ -9,13 +9,21 @@ from localstack.aws.connect import connect_to
 
 LOG = logging.getLogger(__name__)
 
-REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([a-zA-Z0-9_.\-/]+)}}")
+REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([a-zA-Z0-9_.\-/]+)(?::(\d+))?}}")
 
 
 @dataclass
 class DynamicReference:
     service_name: str
     reference_key: str
+    version: str | None = None
+
+    @property
+    def versioned_name(self) -> str:
+        if self.version:
+            return f"{self.reference_key}:{self.version}"
+        else:
+            return self.reference_key
 
 
 def perform_dynamic_reference_lookup(
@@ -30,7 +38,7 @@ def perform_dynamic_reference_lookup(
     if reference.service_name == "ssm":
         ssm_client = connect_to(aws_access_key_id=account_id, region_name=region_name).ssm
         try:
-            return ssm_client.get_parameter(Name=reference.reference_key)["Parameter"]["Value"]
+            return ssm_client.get_parameter(Name=reference.versioned_name)["Parameter"]["Value"]
         except ClientError as e:
             LOG.error("client error accessing SSM parameter '%s': %s", reference.reference_key, e)
             raise

--- a/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
@@ -2,6 +2,7 @@ import json
 import logging
 import re
 from dataclasses import dataclass
+from typing import Any
 
 from botocore.exceptions import ClientError
 
@@ -18,6 +19,13 @@ REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):([^${}]+)}}")
 class DynamicReference:
     service_name: str
     reference_key: str
+
+
+def extract_dynamic_reference(value: Any) -> DynamicReference | None:
+    if isinstance(value, str):
+        if dynamic_ref_match := REGEX_DYNAMIC_REF.search(value):
+            return DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
+    return None
 
 
 def perform_dynamic_reference_lookup(

--- a/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
+++ b/localstack-core/localstack/services/cloudformation/engine/v2/resolving.py
@@ -2,7 +2,6 @@ import json
 import logging
 import re
 from dataclasses import dataclass
-from typing import Any
 
 from botocore.exceptions import ClientError
 
@@ -17,13 +16,6 @@ REGEX_DYNAMIC_REF = re.compile(r"{{resolve:([^:]+):(.+)}}")
 class DynamicReference:
     service_name: str
     reference_key: str
-
-
-def extract_dynamic_reference(value: Any) -> DynamicReference | None:
-    if isinstance(value, str):
-        if dynamic_ref_match := REGEX_DYNAMIC_REF.match(value):
-            return DynamicReference(dynamic_ref_match[1], dynamic_ref_match[2])
-    return None
 
 
 def perform_dynamic_reference_lookup(

--- a/tests/aws/services/cloudformation/test_template_engine.py
+++ b/tests/aws/services/cloudformation/test_template_engine.py
@@ -349,20 +349,20 @@ class TestSsmParameters:
         )
 
     @markers.aws.validated
-    def test_resolve_ssm(self, create_parameter, deploy_cfn_template):
+    def test_resolve_ssm(self, create_parameter, deploy_cfn_template, snapshot):
         parameter_key = f"param-key-{short_uid()}"
         parameter_value = f"param-value-{short_uid()}"
+        snapshot.add_transformer(snapshot.transform.regex(parameter_value, "<parameter-value>"))
         create_parameter(Name=parameter_key, Value=parameter_value, Type="String")
 
         result = deploy_cfn_template(
-            parameters={"DynamicParameter": parameter_key},
+            parameters={"DynamicParameter": parameter_key, "ParameterName": parameter_key},
             template_path=os.path.join(
                 os.path.dirname(__file__), "../../templates/resolve_ssm.yaml"
             ),
         )
 
-        topic_name = result.outputs["TopicName"]
-        assert topic_name == parameter_value
+        snapshot.match("results", result.outputs)
 
     @markers.aws.validated
     def test_resolve_ssm_with_version(self, create_parameter, deploy_cfn_template, aws_client):

--- a/tests/aws/services/cloudformation/test_template_engine.py
+++ b/tests/aws/services/cloudformation/test_template_engine.py
@@ -349,6 +349,7 @@ class TestSsmParameters:
         )
 
     @markers.aws.validated
+    @skip_if_legacy_engine()
     def test_resolve_ssm(self, create_parameter, deploy_cfn_template, snapshot):
         parameter_key = f"param-key-{short_uid()}"
         parameter_value = f"param-value-{short_uid()}"

--- a/tests/aws/services/cloudformation/test_template_engine.py
+++ b/tests/aws/services/cloudformation/test_template_engine.py
@@ -381,8 +381,12 @@ class TestSsmParameters:
             Name=parameter_key, Overwrite=True, Type="String", Value=parameter_value_v2
         )
 
+        versioned_parameter_reference = f"{parameter_key}:{v1['Version']}"
         result = deploy_cfn_template(
-            parameters={"DynamicParameter": f"{parameter_key}:{v1['Version']}"},
+            parameters={
+                "DynamicParameter": versioned_parameter_reference,
+                "ParameterName": versioned_parameter_reference,
+            },
             template_path=os.path.join(
                 os.path.dirname(__file__), "../../templates/resolve_ssm.yaml"
             ),

--- a/tests/aws/services/cloudformation/test_template_engine.py
+++ b/tests/aws/services/cloudformation/test_template_engine.py
@@ -520,7 +520,7 @@ class TestSecretsManagerParameters:
         create_secret(Name=parameter_key, SecretString=parameter_value)
 
         result = deploy_cfn_template(
-            parameters={"DynamicParameter": f"{parameter_key}"},
+            parameters={"DynamicParameter": parameter_key},
             template_path=os.path.join(
                 os.path.dirname(__file__),
                 "../../templates",

--- a/tests/aws/services/cloudformation/test_template_engine.snapshot.json
+++ b/tests/aws/services/cloudformation/test_template_engine.snapshot.json
@@ -691,5 +691,14 @@
     "recorded-content": {
       "error-response": "An error occurred (ValidationError) when calling the CreateChangeSet operation: Parameter InputValue should either have input value or default value"
     }
+  },
+  "tests/aws/services/cloudformation/test_template_engine.py::TestSsmParameters::test_resolve_ssm": {
+    "recorded-date": "24-09-2025, 22:06:32",
+    "recorded-content": {
+      "results": {
+        "ParameterValue": "abc:<parameter-value>",
+        "TopicName": "<parameter-value>"
+      }
+    }
   }
 }

--- a/tests/aws/services/cloudformation/test_template_engine.validation.json
+++ b/tests/aws/services/cloudformation/test_template_engine.validation.json
@@ -110,6 +110,15 @@
       "total": 61.74
     }
   },
+  "tests/aws/services/cloudformation/test_template_engine.py::TestSsmParameters::test_resolve_ssm": {
+    "last_validated_date": "2025-09-24T22:07:22+00:00",
+    "durations_in_seconds": {
+      "setup": 0.95,
+      "call": 10.67,
+      "teardown": 50.0,
+      "total": 61.62
+    }
+  },
   "tests/aws/services/cloudformation/test_template_engine.py::TestSsmParameters::test_resolve_ssm_missing_parameter": {
     "last_validated_date": "2025-08-06T09:34:07+00:00",
     "durations_in_seconds": {

--- a/tests/aws/templates/resolve_ssm.yaml
+++ b/tests/aws/templates/resolve_ssm.yaml
@@ -2,14 +2,25 @@ Parameters:
   DynamicParameter:
     Type: String
 
+  ParameterName:
+    Type: String
+
 Resources:
   topic69831491:
     Type: AWS::SNS::Topic
     Properties:
-      TopicName: !Join [ "", [ "{{resolve:ssm:", !Ref DynamicParameter, "}}"  ] ]
+      TopicName: !Join [ "", [ "{{resolve:ssm:", !Ref DynamicParameter, "}}" ] ]
+  MyParameter:
+    Type: AWS::SSM::Parameter
+    Properties:
+      Type: String
+      Value: !Sub "abc:{{resolve:ssm:${ParameterName}}}"
 Outputs:
   TopicName:
     Value:
       Fn::GetAtt:
         - topic69831491
         - TopicName
+
+  ParameterValue:
+    Value: !GetAtt MyParameter.Value

--- a/tests/unit/services/cloudformation/test_cloudformation.py
+++ b/tests/unit/services/cloudformation/test_cloudformation.py
@@ -1,9 +1,12 @@
+import pytest
+
 from localstack.services.cloudformation.api_utils import is_local_service_url
 from localstack.services.cloudformation.deployment_utils import (
     PLACEHOLDER_AWS_NO_VALUE,
     remove_none_values,
 )
 from localstack.services.cloudformation.engine.template_deployer import order_resources
+from localstack.services.cloudformation.engine.v2.resolving import REGEX_DYNAMIC_REF
 
 
 def test_is_local_service_url():
@@ -60,3 +63,16 @@ def test_order_resources():
     )
 
     assert list(sorted_resources.keys()) == ["A", "B"]
+
+
+class TestDynamicResolving:
+    @pytest.mark.parametrize(
+        "value,matches",
+        [
+            pytest.param("{{resolve:ssm:abc123}}", True, id="basic"),
+            pytest.param("abc:{{resolve:ssm:abc123}}:foo", True, id="with-surrounding"),
+            pytest.param("{{resolve:ssm:${ParameterName}}}", False, id="in-sub"),
+        ],
+    )
+    def test_dynamic_ref_regex_matches(self, value, matches):
+        assert bool(REGEX_DYNAMIC_REF.search(value)) == matches

--- a/tests/unit/services/cloudformation/test_cloudformation.py
+++ b/tests/unit/services/cloudformation/test_cloudformation.py
@@ -69,9 +69,21 @@ class TestDynamicResolving:
     @pytest.mark.parametrize(
         "value,matches",
         [
-            pytest.param("{{resolve:ssm:abc123}}", True, id="basic"),
-            pytest.param("abc:{{resolve:ssm:abc123}}:foo", True, id="with-surrounding"),
-            pytest.param("{{resolve:ssm:${ParameterName}}}", False, id="in-sub"),
+            pytest.param("{{resolve:ssm:abc123}}", True, id="ssm-basic"),
+            pytest.param("abc:{{resolve:ssm:abc123}}:foo", True, id="ssm-with-surrounding"),
+            pytest.param("{{resolve:ssm:${ParameterName}}}", False, id="ssm-in-sub"),
+            pytest.param("{{resolve:secretsmanager:foo}}", True, id="secrets-basic"),
+            pytest.param(
+                "{{resolve:secretsmanager:arn:aws:secretsmanager:us-east-1:000000000000:secret:foo:SecretString:}}",
+                True,
+                id="secrets-partial",
+            ),
+            pytest.param(
+                "{{resolve:secretsmanager:arn:aws:secretsmanager:us-east-1:000000000000:secret:foo:SecretString:::}}",
+                True,
+                id="secrets-full",
+            ),
+            pytest.param("{{resolve:secretsmanager:${SecretName}}}", False, id="secrets-in-sub"),
         ],
     )
     def test_dynamic_ref_regex_matches(self, value, matches):

